lib/LocalServer.js

```diff
@@ -0,0 +1,141 @@
+/**
+    Licensed to the Apache Software Foundation (ASF) under one
+    or more contributor license agreements.  See the NOTICE file
+    distributed with this work for additional information
+    regarding copyright ownership.  The ASF licenses this file
+    to you under the Apache License, Version 2.0 (the
+    "License"); you may not use this file except in compliance
+    with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+*/
+
+var Q = require('q'),
+    io = require('socket.io'),
+    logger = require('./utils').logger,
+    exec = require('./utils').exec,
+    path = require('path'),
+    util = require('util'),
+    portChecker = require('tcp-port-used'),
+    EventEmitter = require('events').EventEmitter;
+
+function LocalServer(port, externalServerUrl) {
+    this.port = port;
+    this.externalServerUrl = externalServerUrl;
+}
+
+util.inherits(LocalServer, EventEmitter);
+
+LocalServer.startServer = function(ports, externalServerUrl, useTunnel) {
+    logger.normal("local-server: scanning ports from " + ports.start + " to " + ports.end);
+
+    return LocalServer.getFirstAvailablePort(ports.start, ports.end)
+        .then(function(port) {
+            logger.normal("local-server: port " + port + " is available");
+            logger.info("local-server: starting local medic server");
+
+            var localServer = new LocalServer(port, externalServerUrl);
+            localServer.createSocketListener();
+
+            if (useTunnel) {
+                return Q.promise(function(resolve) {
+                    localServer.createTunnel().then(function(){
+                        resolve(localServer);
+                    });
+                });
+            }
+
+            return localServer;
+        });
+};
+
+LocalServer.getFirstAvailablePort = function(startPort, endPort) {
+    var ports = Array.apply(null, Array(endPort - startPort + 1)).map(function(element, index) {
+        return startPort + index;
+    });
```

Comment:
Nitpick: it would be more readable to define a `range` function that returns an array of numbers from `start` to `end` and to have `var ports = range(startPort, endPort + 1)`.